### ISSUE TITLE
Intercepted HTTP request gets pushed to the API as soon as another Http request is invoked

### ISSUE THREAD

**Issue by MarinkoSpasojevic** [jsakamoto/Toolbelt.Blazor.HttpClientInterceptor#4]
Hello.

I am not sure if this is an issue or just a default behavior but I am sure you will know more than I do for sure.

My flow is as follows:

1. Navigate to the component that sends an HTTP request to the API
2. Interceptor intercepts that request and inside it, I call another service to check if a token should be refreshed 
3. If it does, I call the RefreshToken method which sends another HTTP request to the API's /token endpoint
4. At this point, the intercepted HTTP request is released as well and it reaches the endpoint before the token gets refreshed

So my question is: why the Interceptor release the intercepted request, even though all other actions are awaited? 

I even tried creating another client with IHttpFactory for the refresh token request, which is not intercepted, but again, as soon as the refresh request is sent, the intercepted one is sent as well.

Thanks for your time. 

Of course, I can share code if required.

Best regards.

**Comment by jsakamoto** [comment]
@MarinkoSpasojevic Thank you for your report. 👍 

Your problem looks like a general problem that is related to async operations.

To resolve this issue, I published the new version of this library, ver.9.2.0 preview 1.

- https://www.nuget.org/packages/Toolbelt.Blazor.HttpClientInterceptor/9.2.0-preview.1

This version provides async version events such as `BeforeSendAsync` and `AfterSendAsync`.

I think, If you upgrade the version of this package to v.9.2.0 preview 1, and handle the event `BeforeSendAsync` instead of the `BeforeSend`, and sends another HTTP request inside the `BeforeSendAsync` event handler, then it will work fine as you expected.

You can also find the sample code in the following link.

- https://github.com/jsakamoto/Toolbelt.Blazor.HttpClientInterceptor/tree/master/SampleSite/ReentryScenarioApp

Could you try it out?

**Comment by MarinkoSpasojevic** [comment]
@jsakamoto I will try it. Just two more questions (just for my curiosity).

In the library description, you mentioned if we use the interceptor in the component it should be disposed of. So my question is should it be disposed of even if we use it as a service injection in another service? 

And, the EnableIntercept method is an extension on top of the HttpClient class and expects a ServiceProvider as a parameter. What about the AddHttpClient method, which comes from Microsoft.Http.Extensions library? There we can extract the object of type HttpClient but there is no ServiceProvider. So, maybe it would be good to enable such implementation as well.

I will ping you back as soon as I try your new update.

Best regards.

**Comment by jsakamoto** [comment]
@MarinkoSpasojevic 

> In the library description, you mentioned if we use the interceptor in the component it should be disposed of.

No, no. I want to say just "You have to unsubscribe event handling when it is no longer needed.". Don't dispose of the interceptor.

Again, if you do this:

```chsarp
interceptor.BeforeSendAsync += MyHandler;
```

at the end of the lifetime of the object that has the `MyHandler`, you have to do this:

```csharp
interceptor.BeforeSendAsync -= MyHandler;
```

If you don't unsubscribe it, the object that has the `MyHandler` method will be never garbage collected. This means it causes memory leaks.

**Comment by jsakamoto** [comment]
@MarinkoSpasojevic 

> And, the EnableIntercept method is an extension on top of the HttpClient class and expects a ServiceProvider as a parameter. What about the AddHttpClient method, which comes from Microsoft.Http.Extensions library? 

Fortunately, the `AddHttpClient` extension method has an overload version that provides `IServiceProvider` object to the arguments of the call back function.

Therefore, you can implement it like this:

```csharp
builder.Services.AddHttpClient("Foo", (serviceProvider, client) =>
{
  client.EnableIntercept(serviceProvider);
  ...              
});
```

**Comment by MarinkoSpasojevic** [comment]
Thank you very much @jsakamoto 

For the first question, I didn't express myself well, I meant exactly what you have explained in your answer.

For the second question, great, I must admit I didn't know that (which is kind an obvious :D ) 

Finally, regarding my problem and your solution, it is working now like a charm. Thanks for that. I worked a lot with Angular apps and having HttpInterceptor is a life saver. So, having the same in the Blazor is even better.

I have no other place to contact you, and our blog https://code-maze.com (if you heard about us) covers .net core, API, angular, blazor topics. So, I would like to have a great article about this library. If you are interested to write maybe yourself about it, we can talk more on twitter (I have followed you there, but have no options to send you a message).

Anyhow, thanks a lot, great job here.
